rpcs3/util/coro.hpp

```diff
@@ -0,0 +1,359 @@
+#pragma once
+
+#if defined(__clang__) && !defined(__cpp_impl_coroutine)
+#define __cpp_impl_coroutine 123
+#endif
+
+#include <coroutine>
```

Comment:
`<coroutine>` in libc++ is available since https://github.com/llvm/llvm-project/commit/2e6ae1d3f2de. For older versions there's `<experimental/coroutine>`.
```c++
In file included from Utilities/File.cpp:12:
rpcs3/util/coro.hpp:7:10: fatal error: 'coroutine' file not found
#include <coroutine>
         ^~~~~~~~~~~
```
libc++ on CI can be upgraded separately from Clang. For example, after applying #11214 + 35a3908e4737 ([build log](https://cirrus-ci.com/task/6290777593610240)) the error changes:
```c++
In file included from Utilities/File.cpp:12:
rpcs3/util/coro.hpp:13:15: error: no member named 'coroutine_traits' in namespace 'std'
        using ::std::coroutine_traits;
              ~~~~~~~^
rpcs3/util/coro.hpp:14:15: error: no member named 'coroutine_handle' in namespace 'std'
        using ::std::coroutine_handle;
              ~~~~~~~^
rpcs3/util/coro.hpp:35:10: error: no template named 'coroutine_handle' in namespace 'std'
                                std::coroutine_handle<> await_suspend(std::coroutine_handle<P> h) noexcept
                                ~~~~~^
rpcs3/util/coro.hpp:35:48: error: no template named 'coroutine_handle' in namespace 'std'
                                std::coroutine_handle<> await_suspend(std::coroutine_handle<P> h) noexcept
                                                                      ~~~~~^
rpcs3/util/coro.hpp:44:9: error: no type named 'suspend_always' in namespace 'std'
                        std::suspend_always initial_suspend() { return {}; }
                        ~~~~~^
rpcs3/util/coro.hpp:49:9: error: no template named 'coroutine_handle' in namespace 'std'
                        std::coroutine_handle<> m_handle{};
                        ~~~~~^
rpcs3/util/coro.hpp:118:9: error: no template named 'coroutine_handle' in namespace 'std'
                        std::coroutine_handle<promise_type> coro;
                        ~~~~~^
rpcs3/util/coro.hpp:125:9: error: no template named 'coroutine_handle' in namespace 'std'
                        std::coroutine_handle<> await_suspend(std::coroutine_handle<> h) noexcept
                        ~~~~~^
rpcs3/util/coro.hpp:125:47: error: no template named 'coroutine_handle' in namespace 'std'
                        std::coroutine_handle<> await_suspend(std::coroutine_handle<> h) noexcept
                                                              ~~~~~^
rpcs3/util/coro.hpp:188:32: error: no template named 'coroutine_handle' in namespace 'std'
                explicit constexpr lazy(std::coroutine_handle<promise_type> h) noexcept
                                        ~~~~~^
rpcs3/util/coro.hpp:193:8: error: no template named 'coroutine_handle' in namespace 'std'
                std::coroutine_handle<promise_type> m_handle;
                ~~~~~^
rpcs3/util/coro.hpp:199:23: error: no member named 'coroutine_handle' in namespace 'std'
                return lazy<T>{std::coroutine_handle<lazy_promise>::from_promise(*this)};
                               ~~~~~^
rpcs3/util/coro.hpp:199:40: error: 'lazy_promise' does not refer to a value
                return lazy<T>{std::coroutine_handle<lazy_promise>::from_promise(*this)};
                                                     ^
rpcs3/util/coro.hpp:53:10: note: declared here
                struct lazy_promise final : lazy_promise_base
                       ^
rpcs3/util/coro.hpp:199:55: error: no member named 'from_promise' in the global namespace
                return lazy<T>{std::coroutine_handle<lazy_promise>::from_promise(*this)};
                                                                  ~~^
rpcs3/util/coro.hpp:205:24: error: no member named 'coroutine_handle' in namespace 'std'
                return lazy<T&>{std::coroutine_handle<lazy_promise>::from_promise(*this)};
                                ~~~~~^
rpcs3/util/coro.hpp:205:41: error: 'lazy_promise' does not refer to a value
                return lazy<T&>{std::coroutine_handle<lazy_promise>::from_promise(*this)};
                                                      ^
rpcs3/util/coro.hpp:80:10: note: declared here
                struct lazy_promise<T&> final : lazy_promise_base
                       ^
rpcs3/util/coro.hpp:205:56: error: no member named 'from_promise' in the global namespace
                return lazy<T&>{std::coroutine_handle<lazy_promise>::from_promise(*this)};
                                                                   ~~^
rpcs3/util/coro.hpp:210:26: error: no member named 'coroutine_handle' in namespace 'std'
                return lazy<void>{std::coroutine_handle<lazy_promise>::from_promise(*this)};
                                  ~~~~~^
rpcs3/util/coro.hpp:210:43: error: 'lazy_promise' does not refer to a value
                return lazy<void>{std::coroutine_handle<lazy_promise>::from_promise(*this)};
                                                        ^
rpcs3/util/coro.hpp:101:10: note: declared here
                struct lazy_promise<void> final : lazy_promise_base
                       ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
```